src/pip/_internal/index.py

```diff
@@ -73,6 +85,87 @@ def _get_content_type(url, session):
     return resp.headers.get("Content-Type", "")
 
 
+class _SchemeIsVCS(Exception):
+    pass
+
+
+class _NotHTML(Exception):
+    pass
+
+
+def _get_html_response(url, filename, session):
+    """Access an HTML page with GET, and return the response.
+
+    This consists of five parts:
+
+    1. Check whether the scheme is supported (non-VCS). Raises `_SchemeIsVCS`
+       on failure.
+    2. Check whether the URL looks like an archive. If it does, make a HEAD
+       request to check the Content-Type header is indeed HTML, and raise
+       `_NotHTML` if it's not. Raise HTTP exceptions on network failures.
+    3. If URL scheme is file: and points to a directory, make it point to
+       index.html instead.
+    4. Actually perform the request. Raise HTTP exceptions on network failures.
+    5. Check whether Content-Type header to make sure the thing we got is HTML,
+       and raise `_NotHTML` if it's not.
+    """
+    # Check for VCS schemes that do not support lookup as web pages.
+    vcs_scheme = _match_vcs_scheme(url)
+    if vcs_scheme:
+        raise _SchemeIsVCS(vcs_scheme)
```

Comment:
Hmm, using exceptions for flow control here seems to me to be adding extra complexity for not much gain. For example, wouldn't it be more straightforward simply to call an `_is_vcs_scheme()` function inside `_get_html_page()` prior to making the request, and then returning `None` as needed? Then you don't need to introduce the one-off `_SchemeIsVCS` as an intermediary. Normally, using exceptions is a last resort thing IMO. I imagine at least some of the other ones can be handled similarly but thought I'd spell out only one.

Reply:
Hmm, I guess it would work. I went for this still because I wanted to distinguish between the two different early returns, so `_get_html_page()` get log them appropriately, but you’re right, the VCS check can be done in `_get_html_page()` instead.

I feel the content type check might be a bit more tricky to handle. The only alternative I can think of is to have two duplicate logging calls.

Reply:
Yes, I think it will become clearer if you start with some of the easier stuff, and then we can see if the duplication is needed or not. I was also wondering if you might actually want two different log messages anyways, since the two code paths are different.